File: plsql-checks/src/main/java/org/sonar/plsqlopen/checks/SameBranchCheck.java
/*
 * Sonar PL/SQL Plugin (Community)
 * Copyright (C) 2015-2017 Felipe Zorzo
 * mailto:felipebzorzo AT gmail DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plsqlopen.checks;

import java.util.ArrayList;
import java.util.List;

import org.sonar.check.Priority;
import org.sonar.check.Rule;
import org.sonar.plugins.plsqlopen.api.PlSqlGrammar;
import org.sonar.plsqlopen.annnotations.ActivatedByDefault;
import org.sonar.plsqlopen.annnotations.ConstantRemediation;
import com.google.common.collect.ImmutableList;
import com.sonar.sslr.api.AstNode;

@Rule(
    key = SameBranchCheck.CHECK_KEY,
    priority = Priority.MAJOR
)
@ConstantRemediation("10min")
@ActivatedByDefault
public class SameBranchCheck extends AbstractBaseCheck {

    public static final String CHECK_KEY = "SameBranch";

    @Override
    public void init() {
        subscribeTo(PlSqlGrammar.IF_STATEMENT);
    }

    @Override
    public void visitNode(AstNode node) {
        List<AstNode> branches = collectStatementsFromBranches(node);
        findSameBranches(branches);
    }
    
    private static List<AstNode> collectStatementsFromBranches(AstNode node) {
        List<AstNode> statementsFromBranches = new ArrayList<>();
        
        statementsFromBranches.add(getStatements(node));
        
        for (AstNode branch : node.getChildren(PlSqlGrammar.ELSIF_CLAUSE)) {
            statementsFromBranches.add(getStatements(branch));
        }
        
        AstNode elseBranch = node.getFirstChild(PlSqlGrammar.ELSE_CLAUSE);
        if (elseBranch != null) {
            statementsFromBranches.add(getStatements(elseBranch));
        }
        
        return statementsFromBranches;
    }
    
    private static AstNode getStatements(AstNode node) {
        return node.getFirstChild(PlSqlGrammar.STATEMENTS);
    }
    
    private void findSameBranches(List<AstNode> branches) {
        for (int i = 1; i < branches.size(); i++) {
            checkBranch(branches, i);
        }
    }
    
    private void checkBranch(List<AstNode> branches, int index) {
        AstNode branch = branches.get(index);
        final int previousBranchIndex = index - 1;
        AstNode otherBranch = branches.get(previousBranchIndex);
        if (CheckUtils.equalNodes(otherBranch, branch)) {
            getContext().createViolation(this, getLocalizedMessage(CHECK_KEY), 
                    branch,
                    ImmutableList.of(newLocation("Original", otherBranch)),
                    otherBranch.getToken().getLine());
            return;
        }
    }

}

File: plsql-checks/src/main/kotlin/org/sonar/plsqlopen/checks/SameBranchCheck.kt
/**
 * Z PL/SQL Analyzer
 * Copyright (C) 2015-2019 Felipe Zorzo
 * mailto:felipebzorzo AT gmail DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plsqlopen.checks

import com.sonar.sslr.api.AstNode
import org.sonar.plsqlopen.asTree
import org.sonar.plsqlopen.sslr.IfStatement
import org.sonar.plugins.plsqlopen.api.PlSqlGrammar
import org.sonar.plugins.plsqlopen.api.annotations.*
import java.util.*

@Rule(key = SameBranchCheck.CHECK_KEY, priority = Priority.MAJOR, tags = [Tags.BUG])
@ConstantRemediation("10min")
@RuleInfo(scope = RuleInfo.Scope.ALL)
@ActivatedByDefault
class SameBranchCheck : AbstractBaseCheck() {

    override fun init() {
        subscribeTo(PlSqlGrammar.IF_STATEMENT)
    }

    override fun visitNode(node: AstNode) {
        val ifStatement = node.asTree<IfStatement>()
        val branches = collectStatementsFromBranches(ifStatement)
        findSameBranches(branches)
    }

    private fun findSameBranches(branches: List<AstNode>) {
        for (i in 1 until branches.size) {
            checkBranch(branches, i)
        }
    }

    private fun checkBranch(branches: List<AstNode>, index: Int) {
        val branch = branches[index]
        val previousBranchIndex = index - 1
        val otherBranch = branches[previousBranchIndex]
        if (CheckUtils.equalNodes(otherBranch, branch)) {
            addIssue(branch, getLocalizedMessage(CHECK_KEY), otherBranch.token.line)
                    .secondary(otherBranch, "Original")
        }
    }

    private fun collectStatementsFromBranches(ifStatement: IfStatement): List<AstNode> {
        val statementsFromBranches = ArrayList<AstNode>()

        statementsFromBranches.add(ifStatement.statements)

        for (branch in ifStatement.elsifClauses) {
            statementsFromBranches.add(branch.statements)
        }

        val elseBranch = ifStatement.elseClause
        if (elseBranch != null) {
            statementsFromBranches.add(elseBranch.statements)
        }

        return statementsFromBranches
    }

    companion object {
        internal const val CHECK_KEY = "SameBranch"
    }

}
